ngraph/frontend/generic/src/plugin_loader.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#ifdef _WIN32
+#include <Windows.h>
+#include <direct.h>
+const char FileSeparator[] = "\\";
+#else  // _WIN32
+
+#include <unistd.h>
+#include <dirent.h>
+#include <dlfcn.h>
+
+const char FileSeparator[] = "/";
+#endif  // _WIN32
+
+#include <string>
+#include <vector>
+#include <sys/stat.h>
+
+#include "plugin_loader.hpp"
+
+using namespace ngraph;
+using namespace ngraph::frontend;
+
+#ifdef WIN32
+#define DLOPEN(fileStr) LoadLibrary(TEXT(fileStr.c_str()))
+#define DLSYM(obj, func) GetProcAddress(obj, func)
+#define DLCLOSE(obj) FreeLibrary(obj)
+#else
+#define DLOPEN(fileStr) dlopen(file.c_str(), RTLD_LAZY)
+#define DLSYM(obj, func) dlsym(obj, func)
+#define DLCLOSE(obj) dlclose(obj)
+#endif
+
+// TODO: change to std::filesystem for C++17
+static std::vector<std::string> listFiles(const std::string& path)
+{
+    std::vector<std::string> res;
+#ifndef WIN32
+    struct dirent *ent;
+    DIR *dir = opendir(path.c_str());
+    if (dir != nullptr)
+    {
+        std::unique_ptr<DIR, void (*)(DIR *)> closeGuard(dir, [](DIR *d)
+        { closedir(d); });
+        while ((ent = readdir(dir)) != NULL)
+        {
+            auto file = path + FileSeparator + std::string(ent->d_name);
+            struct stat stat_path;
+            stat(file.c_str(), &stat_path);
+            if (!S_ISDIR(stat_path.st_mode) && file.find("_frontend.so") != std::string::npos)
+            {
+                res.push_back(std::move(file));
+            }
+        }
+    }
+#else
+    std::string searchPath = path + FileSeparator + "*_frontend.dll";
+    WIN32_FIND_DATA fd;
+    HANDLE handle = ::FindFirstFile(searchPath.c_str(), &fd);
+    if (handle != INVALID_HANDLE_VALUE)
+    {
+        do
+        {
+            if (!(fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) )
+            {
+                res.push_back(path + FileSeparator + fd.cFileName);
+            }
+        } while(::FindNextFile(handle, &fd));
+        ::FindClose(handle);
+    }
+#endif
+    return res;
+}
```

Comment:
looks like we have the same functionality in IE to load plugins. Can we create an object library with `SharedObjectLoader` link to frontends and IE?

Reply:
Approach to have separate library with SharedObjectLoader (located somewhere in common location for IE and Ngraph) looks too complicated comparing to such `#ifdef` wrappers around dlopen/dlclose/dlsym
Rest of functionality is frontend-specific, e.g. look for a directory, find plugins, call methods, cast results, return handles, etc.